src/cmdline/mod.rs

```diff
@@ -90,117 +95,294 @@ fn valid_element(s: &str) -> Result<()> {
+        })
     }
 
-    fn has_capacity(&self, more: usize) -> Result<()> {
-        let needs_space = if self.line.is_empty() { 0 } else { 1 };
-        if self.line.len() + more + needs_space < self.capacity {
+    /// Tries to set a new capacity of the `Cmdline`. It fails when `new_capacity`
+    /// is smaller than the C complatible representation of the current `Cmdline`
+    /// (which requires to be null terminated)
+    ///
+    /// # Arguments
+    ///
+    /// * `new_capacity` - New capacity to be set..
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new(3).unwrap();
+    /// assert!(cl.insert_string("foo", CmdlineArgType::BootArg).is_err());
+    /// cl.set_capacity(4).unwrap();
+    /// assert!(cl.insert_string("foo", CmdlineArgType::BootArg).is_ok());
+    /// assert_eq!(cl.set_capacity(3), Err(Error::InvalidCapacity));
+    /// ```
+    pub fn set_capacity(&mut self, new_capacity: usize) -> Result<()> {
+        if self.get_null_terminated_representation_size() <= new_capacity {
+            self.capacity = new_capacity;
+
             Ok(())
         } else {
-            Err(Error::TooLarge)
+            Err(Error::InvalidCapacity)
         }
     }
 
-    fn start_push(&mut self) {
-        if !self.line.is_empty() {
-            self.line.push(' ');
+    fn get_null_terminated_representation_size(&self) -> usize {
+        // Computing current size of the cmdline
+        let mut cmdline_size = self.boot_args.len() + 1; // for null terminator
+
+        if !self.init_args.is_empty() {
+            cmdline_size += INIT_ARGS_SEPARATOR.len() + self.init_args.len();
         }
+
+        cmdline_size
     }
 
-    fn end_push(&mut self) {
-        // This assert is always true because of the `has_capacity` check that each insert method
-        // uses.
-        assert!(self.line.len() < self.capacity);
+    fn has_capacity(&self, more: usize, args_type: CmdlineArgType) -> bool {
+        let mut cmdline_size = self.get_null_terminated_representation_size();
+
+        // Adding extra size required for the insertion of `more` bytes as a `args_type` argument
+
+        // Checking if one extra space needs to be added before insertion of a new argument
+        cmdline_size += match args_type {
+            CmdlineArgType::BootArg => {
+                if !self.boot_args.is_empty() {
+                    1
+                } else {
+                    0
+                }
+            }
+            CmdlineArgType::InitArg => {
+                if !self.init_args.is_empty() {
+                    1
+                } else {
+                    INIT_ARGS_SEPARATOR.len()
+                }
+            }
+        };
+
+        cmdline_size += more;
+
+        cmdline_size <= self.capacity
     }
 
-    /// Validates and inserts a key-value pair into this command line.
+    /// Validates and inserts a key=value pair as a new cmdline argument.
```

Comment:
Why change key=value? The `key-value` is written like that on purpose as it represents a concept. In the documentation comments we should use concepts, not implementation details as a general rule.

Reply:
keeping as it was initially